### PR TITLE
Synchronise expiry of id token cookie with JWT exp claim

### LINKED ISSUE #32485: oauth2: ID Token JWT expiry may not be synchronised with IdToken cookie expiry
*Title*: The expiry of the IdToken cookie may be longer than the expiry of the JWT stored inside the cookie

*Description*:
This issue is related to the oauth2 and jwt filters. The authn server sets the `expires_in` field of the token response in sync with the expiry of the access token. The oauth2 filter however uses this value to set the expiry of all cookies, including the IdToken cookie. However, authn providers are not obliged to use the same lifetime for access tokens and ID tokens. For example, Microsoft Entra does not: https://learn.microsoft.com/en-us/entra/identity-platform/configurable-token-lifetimes. By default, Entra ID tokens have a lifetime of 60 minutes, whereas Entra access tokens have a lifetime of somewhere in between 60 and 90 minutes. The effect of this is that the IdToken cookie expires sometime after the JWT contained in it. As a result, after 60 minutes pass, the JWT filter errors with `Jwt is expired` and the user must delete the cookies or wait until the cookie expires as well.

This is related to https://github.com/envoyproxy/envoy/issues/30053, because the root cause of both issues is the same: the `expires_in` field of the token response must only be used to determine expiry of the access token, not of any other tokens. The fix should also roughly be the same for both issues: if the IdToken cookie expiry is set from the JWT expiry field, the problem is solved.

*Repro steps*:
Configure your OAuth2 authorization server to hand out access tokens and ID tokens with different lifetime. In envoy, make use of the ID token in the jwt filter, e.g. like this:
```
from_cookies:
  - IdToken
forward_payload_header: X-Id-Token-Payload
```

After the ID token expires, envoy responds with `Jwt is expired` until the cookies / access token expire as well.

*Other comments*:
Another way to solve the issue would be to introduce a separate configuration field, named e.g. `max_expiry`. Say this is set to 60 minutes, then all cookies would not have an expiry of more than 60 minutes. This would fix the issue as well, although it sounds a bit like a hack to me. In addition, `max_expiry` would have to exclude the refresh token, as otherwise https://github.com/envoyproxy/envoy/issues/30053 would happen again.

### PR DESCRIPTION
**Pull request by michaelsauter**
The expiry of the id token cookie should match the exp claim in the id token JWT, not the expiry of the access token. Some identity providers (definitely Microsoft Entra ID, probably others) issue id tokens with different expiry than the access tokens.

This change is largely based on the work done in #32278 which supports expiry of the refresh token based on the exp claim of the refresh token if it is a JWT.

Closes #32485.

Note: #32278 changed the httpOnly setting of the refresh token cookie without discussion around it. I decided to "revert" that change so that the httpOnly is controlled again by the runtime guard. Let me know if you would like to change this and remove the runtime guard completely? None of these cookies should be accessible by the client ...

### CODE REVIEW

**Comment by michaelsauter** [comment]
> This needs a clang-format (if you click through a few things from the "prechecks precheck format" details you get to [here](https://dev.azure.com/cncf/envoy/_build/results?buildId=170382&view=logs&j=c7de00aa-b0fb-52ba-707b-da6afd918eda&t=449299ab-9da3-5c6b-afe3-e0c5dbf1dfd1&l=153) eventually which provides a diff you could apply).

Yes sorry, realised after submitting the PR. I did not get to update it yesterday but will do so today.

**Comment by michaelsauter** [comment]
This should be ready for review now.

File: source/extensions/filters/http/oauth2/filter.cc [hunk 1]
```diff
@@ -577,19 +578,45 @@ OAuth2Filter::getExpiresTimeForRefreshToken(const std::string& refresh_token,
...
   }
   return std::to_string(expires_in.count());
 }
 
+std::string OAuth2Filter::getExpiresTimeForIdToken(const std::string& id_token,
+                                                   const std::chrono::seconds& expires_in) const {
+  if (!id_token.empty()) {
+    ::google::jwt_verify::Jwt jwt;
+    if (jwt.parseFromString(id_token) == ::google::jwt_verify::Status::Ok && jwt.exp_ != 0) {
+      const std::chrono::seconds expirationFromJwt = std::chrono::seconds{jwt.exp_};
```
Comment (lizan) [inline source/extensions/filters/http/oauth2/filter.cc:599]: style nit: expiration_from_jwt

File: test/extensions/filters/http/oauth2/filter_test.cc [hunk 2]
```diff
@@ -2049,6 +2049,200 @@ TEST_P(OAuth2Test, OAuthAccessTokenSucessWithTokensUseRefreshTokenAndNoExpClaimI
...
+
+  // host_ must be set, which is guaranteed (ASAN).
+  Http::TestRequestHeaderMapImpl request_headers{
+      {Http::Headers::get().Host.get(), "traffic.example.com"},
+      {Http::Headers::get().Path.get(), "/_signout"},
+      {Http::Headers::get().Method.get(), Http::Headers::get().MethodValues.Get},
+  };
+  filter_->decodeHeaders(request_headers, false);
+
+  const std::string idToken =
```
Comment (lizan) [inline test/extensions/filters/http/oauth2/filter_test.cc:2219]: id_token

**Review by lizan — COMMENTED** [review]
Sorry about the delay. just few nits and this is good to go. (approved CI run and it passes)

**Review by lizan — APPROVED** [review]

**Comment by michaelsauter** [comment]
@lizan Thanks for approving! I just resolved the merge conflict on the changelog, this should be ready to merge now.

**Review by soulxu — APPROVED** [review]
LGTM, thanks!

@lizan already approve the PR, I will merge it when the CI happy

**Comment by soulxu** [comment]
@michaelsauter maybe worth to try to merge with main, then see if the CI is happy or not.

**Comment by tyxia** [comment]
@michaelsauter  Please merge main to resolve the conflicts.

I will merge it once that is done and CI is happy. Thanks!

**Comment by michaelsauter** [comment]
@soulxu @tyxia Sorry finally got around to look at this again. I merged main (quite some changes ...) and hope that CI passes now (I wasn't able to run everything locally atm).

**Comment by michaelsauter** [comment]
@tyxia Merged & pushed

**Review by tyxia — APPROVED** [review]
Thanks!

@lizan @soulxu  has approved

**Pull request merged**